src/apps/packet_filter/conntrack.lua

```diff
@@ -72,229 +69,195 @@ ffi.cdef [[
    } __attribute__((packed)) conn_spec_ipv6;
 ]]
 
-local conn_spec_ipv4 = ffi.typeof 'conn_spec_ipv4'
-
-
-local function conn_spec_from_ipv4_header(b)
-   local spec = conn_spec_ipv4()
-   local flags = 0
-   do
-      local hdr_ips = ffi.cast('uint32_t*', b+IPV4_SOURCE_OFFSET)
-      spec.src_ip = hdr_ips[0]
-      spec.dst_ip = hdr_ips[1]
+----
+
+local conntrack
+do
+   local conntracks = {}
+   local counts = {}
+   local time = engine.now
+   local function new() return {{}, {}, time()} end
+   local function put(p, k, v)
+      local isnew = p[1][k] == nil
+      p[1][k] = v
+      return isnew
    end
-   spec.protocol = b[IPV4_PROTOCOL_OFFSET]
-   if spec.protocol == IP_TCP or spec.protocol == IP_UDP then
-      local hdr_ports = ffi.cast('uint16_t*', b+IPV4_SOURCE_PORT_OFFSET)
-      spec.src_port = hdr_ports[0]
-      spec.dst_port = hdr_ports[1]
-      if spec.protocol == IP_TCP then
-         flags = b[IPV4_TCP_FLAGS]
+   local function get(p, k) return p[1][k] or p[2][k] end
+   local function age(p, t)
+      if time() > p[3]+t then
+         p[1], p[2], p[3] = {}, p[1], time()
+         return true
       end
-   else
-      spec.src_port, spec.dst_port = 0, 0
    end
-   return spec, flags
+
+   conntrack = {
+      define = function (name)
+         if not name then return end
+         conntracks[name] = conntracks[name] or new()
+         counts[name] = counts[name] or 0
+      end,
+
+      track = function (name, key, revkey, limit)
+         limit = limit or 1000
+         local p = conntracks[name]
+         if counts[name] > limit then
+            if age(p, 0.01) then
+               counts[name] = 0
+            end
+         end
+         if put(p, key, true) then
+            counts[name] = counts[name] + 1
+         end
+         put(p, revkey, true)
+      end,
+
+      check = function (name, key)
+         return key and get(conntracks[name], key)
+      end,
+
+      age = function(name, t)
+         if name and age(conntracks[name], t or 7200) then
```

Comment:
These are the "tuning" factors.  Should they be configurable?  It's not hard to imagine a setup with several filters, most of them using the defaults (2hours/1000connections/10msec), and one doing some very application-specific filtering and needing different vaules (say, (1min/100conections/1sec)

But you're right, at the very least they should be well-commented constants at the top of the file.